python/mxnet/gluon/block.py

```diff
@@ -954,6 +955,22 @@ def _build_cache(self, *args):
             unused = ', '.join(list(param_names - set(used_param_names)))
             warnings.warn("Parameter %s is not used by any computation. "
                           "Is this intended?"%unused, stacklevel=4)
+        if self._backend:
+            arg_array = []
+            ctx = args[0].context
+            try:
+                for name in out.list_arguments():
+                    if name in data_names.keys():
+                        arg_array.append(args[data_names[name]])
+                    else:
+                        arg_array.append(params.get(name).data())
+
+            except DeferredInitializationError:
```

Comment:
Per offline discussion with @leezu, Gluon should always give shapes as long as we do this:
https://github.com/apache/incubator-mxnet/blob/93b72c03d9871b71fa274b3c34aa6a4473be2155/python/mxnet/gluon/block.py#L1032-L1040

Reply:
Ie. instead of handling except DeferredInitializationError as `arg_array = None`, you need to infer the parameter shapes earlier on based on the symbol, before creating the CachedOp

Reply:
@leezu This is not always guaranteed to work. The input shapes may not be enough to infer shapes through the model. And the shapes inferred may not the be the same as the actual shapes used at inference time. 

Some backends may not be ok with this. They may error out saying the shapes presented at partition time do not match those at runtime. So the safe approach is either to provide the exact shapes at partition time, or not at all.

Reply:
Plus, during training we dont have the params yet (shapes/types) so we need to partition without them. For inference we do have the shapes/types, so it makes sense to have a flow that works for both use-cases.

The question is how we enable this flow with Gluon-style handling (ie. just make it work internally by catching the exception, or have the user pass in a flag to explicitly skip the attempt to get the params)

Reply:
DeferredInitializationError is with respect to parameter shapes.

> The input shapes may not be enough to infer shapes through the model.

If the input shapes are not sufficient to infer parameter shapes, forward will not be possible.
It's correct that dynamic shape operators prevent shape inference of the complete model.

My concern here is only to handle correctly the case where parameter shape can be inferred correctly and the PR currently does not. Let's discuss more about the complete model shape inference.

Reply:
Thanks again @leezu! made the code changes based on our offline discussion!